lightning/src/util/ser_macros.rs

```diff
@@ -372,7 +372,87 @@ macro_rules! impl_writeable_tlv_based {
 				}, {
 					$(($vectype, $vecfield)),*
 				});
-				_init_tlv_based_struct!({$($reqfield),*}, {$($field),*}, {$($vecfield),*})
+				_init_tlv_based_struct!($st, {$($reqfield),*}, {$($field),*}, {$($vecfield),*})
+			}
+		}
+	}
+}
+
+/// Implement Readable and Writeable for an enum, with struct variants stored as TLVs and tuple
+/// variants stored directly.
+/// The format is, for example
+/// impl_writeable_tlv_based_enum!(EnumName,
+///   (0, VariantName) => { (0, variant_field)}, {(1, variant_optional_field)}, {};
+///   (1, TupleVariant), (2, TupleVariantB),
+/// );
+/// The type is written as a single byte, followed by any variant data.
+/// Attempts to read an unknown type byte result in DecodeError::UnknownRequiredFeature.
+macro_rules! impl_writeable_tlv_based_enum {
+	($st: ident, $(($variant_id: expr, $variant_name: ident) =>
+		{$(($reqtype: expr, $reqfield: ident)),* $(,)*},
+		{$(($type: expr, $field: ident)),* $(,)*},
+		{$(($vectype: expr, $vecfield: ident)),* $(,)*}
+	),* $(,)*;
+	$(($tuple_variant_id: expr, $tuple_variant_name: ident)),*  $(,)*) => {
+		impl ::util::ser::Writeable for $st {
+			fn write<W: ::util::ser::Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+				match self {
+					$($st::$variant_name { $(ref $reqfield),* $(ref $field),*, $(ref $vecfield),* } => {
+						let id: u8 = $variant_id;
+						id.write(writer)?;
+						_write_tlv_fields!(writer, {
+							$(($reqtype, $reqfield)),*
+						}, {
+							$(($type, $field)),*
+						}, {
+							$(($vectype, Some(::util::ser::VecWriteWrapper(&$vecfield)))),*
+						});
+					}),*
+					$($st::$tuple_variant_name (ref field) => {
+						let id: u8 = $tuple_variant_id;
+						id.write(writer)?;
+						field.write(writer)?;
+					}),*
+				}
+				Ok(())
+			}
+		}
+
+		impl ::util::ser::Readable for $st {
+			fn read<R: ::std::io::Read>(reader: &mut R) -> Result<Self, ::ln::msgs::DecodeError> {
+				let id: u8 = ::util::ser::Readable::read(reader)?;
+				match id {
+					$($variant_id => {
+						// Because read_tlv_fields creates a labeled loop, we cannot call it twice
+						// in the same function body. Instead, we define a closure and call it.
```

Comment:
Within the closure, the labeled loop is called multiple times because the macro gets recursively called on objects within objects, right? 

This comment only made sense after commenting out the closure part and seeing the warning :thinking: So if the above statement is correct, could be helpful to add something to its effect.

Reply:
The macro itself isn't recursive (though we call `read` recursively, of course), but Rust, somewhat obnoxiously, warns you about any two loop labels within one function, even if they aren't in the same scope. eg the following generates completely spurious warnings that you cannot turn off:
```
match a {
    1 => {
        'tlv_loop: loop {}
    },
    2 => {
        'tlv_loop: loop {}
    },
```